pkg/source/hdfsprotocol/hdfs_source_client.go

```diff
@@ -19,54 +19,247 @@ package hdfsprotocol
+	if err != nil {
+		hdfsFile.Close()
+		return nil, nil, err
+	}
+
+	return newHdfsFileReaderClose(hdfsFile, int64(httpRange.EndIndex-httpRange.StartIndex), hdfsFile), source.ResponseHeader{
+		source.LastModified: fileInfo.ModTime().Format(layout),
+	}, nil
+}
+
+func (h *hdfsSourceClient) GetLastModifiedMillis(ctx context.Context, url string, header source.RequestHeader) (int64, error) {
+
+	hdfsClient, path, err := h.getHDFSClientAndPath(url)
+	if err != nil {
+		return -1, err
+	}
+
+	info, err := hdfsClient.Stat(path)
+	if err != nil {
+		return -1, err
+	}
+
+	return info.ModTime().UnixNano() / time.Millisecond.Nanoseconds(), nil
+}
+
+// getHDFSClient return hdfs client
+func (h *hdfsSourceClient) getHDFSClient(rawurl string) (*hdfs.Client, error) {
+	if len(rawurl) < 4 {
+		return nil, errors.Errorf("hdfs url invalid: url is %s", rawurl)
+	}
+
+	parse, err := url.Parse(rawurl)
+	if err != nil {
+		return nil, err
+	}
+
+	// get client for map
+	h.RWMutex.RLock()
+	if client, ok := h.clientMap[parse.Host]; ok {
+		h.RWMutex.RUnlock()
+		return client, nil
+	}
+	h.RWMutex.RUnlock()
+
+	// create client option
+	options := hdfs.ClientOptionsFromConf(map[string]string{
+		"dfs.client.use.datanode.hostname": "true",
+	})
+	options.Addresses = strings.Split(parse.Host, ",")
+	u, err := user.Current()
+	if err != nil {
+		return nil, err
+	}
+	options.User = u.Username
+
+	// create hdfs client and put map
+	h.RWMutex.Lock()
+	client, err := hdfs.NewClient(options)
+	if err != nil {
+		h.RWMutex.Unlock()
+		return nil, err
+	}
+	h.clientMap[parse.Host] = client
+	h.RWMutex.Unlock()
+	return client, err
 }
 
-func (h hdfsSourceClient) GetLastModifiedMillis(ctx context.Context, url string, header source.RequestHeader) (int64, error) {
-	panic("implement me")
+// getHDFSPath return file path
+func (h *hdfsSourceClient) getHDFSPath(urls string) (string, error) {
+	parse, err := url.Parse(urls)
+	if err != nil {
+		return "", err
+	}
+	return parse.Path, nil
+}
+
+// getHDFSClientAndPath return client and path
+func (h *hdfsSourceClient) getHDFSClientAndPath(urls string) (*hdfs.Client, string, error) {
+	client, err := h.getHDFSClient(urls)
+	if err != nil {
+		return nil, "", errors.Errorf("hdfs create client fail, url is %s", urls)
+	}
+	path, err := h.getHDFSPath(urls)
+	if err != nil {
+		return client, "", errors.Errorf("hdfs url path parse fail, url is %s", urls)
+	}
+	return client, path, nil
 }
 
 func NewHDFSSourceClient(opts ...HDFSSourceClientOption) source.ResourceClient {
-	sourceClient := &hdfsSourceClient{}
+	sourceClient := &hdfsSourceClient{
+		clientMap: make(map[string]*hdfs.Client),
+	}
 	for i := range opts {
 		opts[i](sourceClient)
 	}
 	return sourceClient
 }
 
 var _ source.ResourceClient = (*hdfsSourceClient)(nil)
+
+func (rc *hdfsFileReaderClose) Read(p []byte) (n int, err error) {
+	return rc.limited.Read(p)
+}
+
+func (rc *hdfsFileReaderClose) Close() error {
+	return rc.c.Close()
+}
```

Comment:
TODO

Implement `io.WriteTo` for zero copy.

Reply:
Wait for Code Receive.